core/templates/services/exploration-html-formatter.service.ts

```diff
@@ -79,14 +79,12 @@ export class ExplorationHtmlFormatterService {
       parentHasLastAnswerProperty: boolean,
       labelForFocusTarget: string,
       savedSolution: string | null): string {
-    var htmlInteractionId = this.camelCaseToHyphens.transform(interactionId);
-    var element = $('<oppia-interactive-' + htmlInteractionId + '>');
-
-    element = (
+    const htmlInteractionId = this.camelCaseToHyphens.transform(interactionId);
+    const interactionInitTag = '<oppia-interactive-' + htmlInteractionId + '>';
+    let directiveOuterHtml = (
       this.extensionTagAssembler.formatCustomizationArgAttrs(
-        element, interactionCustomizationArgs));
-    const tagEnd = '></oppia-interactive-' + htmlInteractionId + '>';
-    let directiveOuterHtml = element.get(0).outerHTML.replace(tagEnd, '');
+        interactionInitTag, interactionCustomizationArgs)
+    ).replace('>', '');
```

Comment:
Woah -- don't do all this stuff. Never assemble html using strings, it leads to XSS vulns. That's why we were using jquery in the first place.

I'm open to other safe ways to do it but direct string manipulation is best avoided.

(Admittedly the previous code wasn't super-great either, in terms of the '+' concatenation.)

Reply:
Yeah, I agree but I'm not sure how to do it, except for jQuery, which is a dependency that we want to get rid of after we complete the Angular migration.

Also, this whole function does HTML string manipulation, only this small part is handled by jQuery. 

@srijanreddy98 Any idea how to do this using Angular?

Reply:
Couldn't you keep the previous approach as-is for now?

Reply:
As, the only problem is typescript I think it is fine to revert the code and add type ignores together with a TODO to refactor this. The refactor should have a high priority as this is a security issue even with the previous code.

Reply:
TBH, I don't see major security issues here that can't be patched. The most secure version of this would have two changes. There is one variable thats completely unknown (savedSolution) and another (interactionId) that has a limited set of correct values.

For savedSolution:
I looked at the places from where we were calling the getInteractionHtml function. All but one used null. The only exception I saw was in add-or-update-solution-modal-controller where we called it with the string `savedMemento()`.
So we can change the type from `string | null` to `'savedMemento()` | null` and add an if check to verify if thats the case or not.

For inteactionId, we can just check if it is from the list of known interactionIds and throw and error if not. One thing to note is that this won't be the case when we migrate the interaction view to angular. We would be creating is dynamically and not string manipulation would be required. Here is an example of how it would look like. This is from the draft pr for schema-based-editors migration where we are using dynamic component creation to show object editors. Previously it also used strings for the same purpose. https://github.com/oppia/oppia/blob/84eef670a31e061429780719793e1b82b5afc939/core/templates/components/forms/custom-forms-directives/object-editor.directive.ts#L205

Reply:
I also noticed labelForFocusTarget, we can just check if the string only has a-z and A-Z values. That should be a good enough check.

Reply:
I have a feeling that Sean might not be very happy with the suggested solution so here is another one:
<img width="765" alt="Screenshot 2021-11-25 at 20 19 03" src="https://user-images.githubusercontent.com/23002114/143496671-5d391236-1546-409a-b89a-84daafd643fb.png">

Reply:
I do like @srijanreddy98's last solution better :-) Thanks!

One question -- is it possible to easily run the final HTML through a sanitizer before passing it to the template? If not, it's OK, but I just wanted to check.

Reply:
Note: the use of document.createElement() etc. is also supported by e.g. [this security reference](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/DOM_based_XSS_Prevention_Cheat_Sheet.md#guideline-3---use-documentcreateelement-elementsetattributevalue-elementappendchild-and-similar-to-build-dynamic-interfaces).

One other suggestion I have is to add comments above any variables that are used in setAttribute to explain why they're safe. If a developer then makes changes to this part of the file in the future, hopefully they (and the reviewer) will read that and realize that this is a concern to keep in mind.

Reply:
Thanks, Srijan, I think I will go with reverting for now and creating an issue and then I or someone else can implement this according to your recommendations.

Reply:
Done, removed the changes for now.

Reply:
Okay, I decided to implement this right away, the only issue that I face now is that `setAttribute` doesn't allow `[` and `]` in the attribute name (see https://dom.spec.whatwg.org/#ref-for-dfn-throw%E2%91%A3%E2%91%A7 and https://www.w3.org/TR/xml/#NT-Name). 

![image](https://user-images.githubusercontent.com/10142938/144124205-b6a4eae6-1afe-41fb-9a99-abc092650ca0.png)


@srijanreddy98 Do you know any way on how to get around this?

Reply:
It is a bit of a workaround. This throws an exception. So we first add and remove the attribute (without the `[]`). This confirms that the only thing out of line will be `[]`. Then we create the attribute string ourselves and attach it. This might sound complex and it is. Here are some other things to keep in mind:
1. First add all the attributes that don't need the `[]`. Then get the outerHTML. Then remove the `></the-tag>` from the end.
2. Use the safety checks of the setAttribute. If it throws and error, return/ throw error. Else append the attribute string (with the `[]`) using string concatenation.
3. Append the `></the-tag>` after all the attributes have been added and return the string.